src/coreclr/tools/aot/crossgen2/Program.cs

```diff
@@ -529,8 +540,20 @@ private void RunSingleCompilation(Dictionary<string, string> inFilePaths, Instru
             //
             // Initialize output filename
             //
-            var suffixStr = _commandLineOptions.SingleFileCompilation && _commandLineOptions.InputBubble ? ".ni.dll.tmp" : ".ni.dll";
-            var outFile = _commandLineOptions.OutNearInput ? inFilePaths.First().Value.Replace(".dll", suffixStr) : _commandLineOptions.OutputFilePath;
+            string inFilePath = inFilePaths.First().Value;
+            string nearOutFilePath = Path.GetExtension(inFilePath) switch
+            {
+                ".dll" => Path.ChangeExtension(inFilePath,
+                    _commandLineOptions.SingleFileCompilation && _commandLineOptions.InputBubble
+                        ? ".ni.dll.tmp"
+                        : ".ni.dll"),
+                ".exe" => Path.ChangeExtension(inFilePath,
+                    _commandLineOptions.SingleFileCompilation && _commandLineOptions.InputBubble
+                        ? ".ni.exe.tmp"
+                        : ".ni.exe"),
+                _ => inFilePath
```

Comment:
Actually, I think this case should be fully forbidden. It would be incorrect if inputbubble is used, because all ni files should be moved to target dir all at once. Besides, I think I've seen problems when same file is passed as input and output even without inputbubble.

Reply:
Throw an exception is a good idea. I will change it.